File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
const kthSmallest = function(root, k) {
    const values = [];
    const dfs = node => {
        if (!node || values.length === k) return;
        dfs(node.left);
        values.push(node.val);
        dfs(node.right);
    }
    dfs(root);
    return values[k-1]
};

File: medium/230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    data = []
    
    dfs = -> (node, data, k) {
        return if data.size == k || !node
        
        dfs.call node.left, data, k
        data << node.val
        dfs.call node.right, data, k
    }
    
    dfs.call root, data, k
    
    data[k-1]
end